source/extensions/network/dns_resolver/getaddrinfo/getaddrinfo.cc

```diff
@@ -0,0 +1,237 @@
+#include "source/extensions/network/dns_resolver/getaddrinfo/getaddrinfo.h"
+
+#include "envoy/extensions/network/dns_resolver/getaddrinfo/v3/getaddrinfo_dns_resolver.pb.h"
+#include "envoy/network/dns_resolver.h"
+
+#include "source/common/api/os_sys_calls_impl.h"
+#include "source/common/network/address_impl.h"
+
+namespace Envoy {
+namespace Network {
+
+// This resolver uses getaddrinfo() on a dedicated resolution thread. Thus, it is only suitable
+// currently for relatively low rate resolutions. In the future a thread pool could be added if
+// desired.
+class GetAddrInfoDnsResolver : public DnsResolver, public Logger::Loggable<Logger::Id::dns> {
+public:
+  GetAddrInfoDnsResolver(Event::Dispatcher& dispatcher, Api::Api& api)
+      : dispatcher_(dispatcher),
+        resolver_thread_(api.threadFactory().createThread([this] { resolveThreadRoutine(); })) {}
+
+  ~GetAddrInfoDnsResolver() override {
+    {
+      absl::MutexLock guard(&mutex_);
+      shutting_down_ = true;
+    }
+
+    resolver_thread_->join();
+  }
+
+  // DnsResolver
+  ActiveDnsQuery* resolve(const std::string& dns_name, DnsLookupFamily dns_lookup_family,
+                          ResolveCb callback) override {
+    ENVOY_LOG(debug, "adding new query [{}] to pending queries", dns_name);
+    auto new_query = std::make_unique<PendingQuery>(dns_name, dns_lookup_family, callback);
+    absl::MutexLock guard(&mutex_);
+    pending_queries_.emplace_back(std::move(new_query));
+    return pending_queries_.back().get();
+  }
+
+  void resetNetworking() override {}
+
+private:
+  class PendingQuery : public ActiveDnsQuery {
+  public:
+    PendingQuery(const std::string& dns_name, DnsLookupFamily dns_lookup_family, ResolveCb callback)
+        : dns_name_(dns_name), dns_lookup_family_(dns_lookup_family), callback_(callback) {}
+
+    void cancel(CancelReason) override {
+      ENVOY_LOG(debug, "cancelling query [{}]", dns_name_);
+      cancelled_ = true;
+    }
+
+    const std::string dns_name_;
+    const DnsLookupFamily dns_lookup_family_;
+    ResolveCb callback_;
+    bool cancelled_{false};
+  };
+  // Must be a shared_ptr for passing around via post.
+  using PendingQuerySharedPtr = std::shared_ptr<PendingQuery>;
+
+  // RAII wrapper to free the `addrinfo`.
+  class AddrInfoWrapper : NonCopyable {
+  public:
+    AddrInfoWrapper(addrinfo* info) : info_(info) {}
+    ~AddrInfoWrapper() { Api::OsSysCallsSingleton::get().freeaddrinfo(info_); }
+    const addrinfo* get() { return info_; }
+
+  private:
+    addrinfo* info_;
+  };
+
+  // Parse a GAI response and determine the final address list. We could potentially avoid adding
+  // v4 or v6 addresses if we know they will never be used. Right now the final filtering is done
+  // below and this code is kept simple.
+  std::pair<ResolutionStatus, std::list<DnsResponse>>
+  processGaiResponse(const PendingQuery& query, const addrinfo* addrinfo_result) {
```

Comment:
I'd be inclined to remove "Gai" from this method name as the style guide is not a big fan of abbreviations (and since there aren't other kinds of responses which need to be processed)